### ISSUE TITLE
Ruby add_extension requires the addon ID to be provided in the manifest, but does not honour it

### ISSUE THREAD

**Issue by theDogOfPavlov** [SeleniumHQ/selenium#7151]
## 🐛 Bug Report

Ruby add_extension requires the addon ID to be provided in the manifest, but does not honour it.

related issues:
https://github.com/SeleniumHQ/selenium/pull/6550
https://github.com/SeleniumHQ/selenium/issues/6471

## To Reproduce

When adding an extension directly to firefox, if the addon ID is not provided in the manifest, then a temporary one is created and assigned.

Detailed steps to reproduce the behaviour in Ruby Selenium:

If the addon ID is not provided then an error is generated:
```
1556528802974	addons.xpi	DEBUG	New add-on injector@1.0.0 in app-profile
1556528803006	addons.xpi-utils	DEBUG	New add-on injector@1.0.0 installed in app-profile
1556528803091	addons.xpi-utils	WARN	addMetadata: Add-on injector@1.0.0 is invalid: Error: Invalid addon ID: expected addon ID injector@1.0.0, found undefined in manifest(resource://gre/modules/addons/XPIDatabase.jsm:2357:15) JS Stack trace: addMetadata@XPIDatabase.jsm:2357:15
```

If the addon ID is provided (in the recommended format) then an error is generated:
```
1556532350039	addons.xpi	DEBUG	New add-on injector@1.0.0 in app-profile
1556532350071	addons.xpi-utils	DEBUG	New add-on injector@1.0.0 installed in app-profile
1556532350160	addons.xpi-utils	WARN	addMetadata: Add-on injector@1.0.0 is invalid: Error: Invalid addon ID: expected addon ID injector@1.0.0, found injector@injected.com in manifest(resource://gre/modules/addons/XPIDatabase.jsm:2357:15) JS Stack trace: addMetadata@XPIDatabase.jsm:2357:15 
```

If the addon ID is provided in the format preferred by selenium (name@version) then the extension is loaded:
```
1556532773570	addons.xpi	DEBUG	New add-on injector@1.0.0 in app-profile
1556532773601	addons.xpi-utils	DEBUG	New add-on injector@1.0.0 installed in app-profile
1556532773767	addons.manager	DEBUG	Registering startup change 'installed' for injector@1.0.0
1556532773770	addons.xpi	DEBUG	Loading bootstrap scope from file:///tmp/rust_mozprofile.awdd2tQ6RVlm/extensions/injector@1.0.0/
1556532773770	addons.xpi	DEBUG	Calling bootstrap method install on injector@1.0.0 version 1.0.0
1556532773770	addons.xpi-utils	DEBUG	Make addon app-profile:injector@1.0.0 visible
1556532773770	addons.xpi	DEBUG	Updating XPIState for {"id":"injector@1.0.0","syncGUID":"{bb2801d2-9ef5-44b4-a4a1-e4e3f5879bce}","version":"1.0.0","type":"extension","loader":null,"updateURL":null,"optionsURL":null,"optionsType":null,"optionsBrowserStyle":true,"aboutURL":null,"defaultLocale":{"name":"injector","creator":null,"developers":null,"translators":null,"contributors":null},"visible":true,"active":false,"userDisabled":false,"appDisabled":false,"installDate":1556532763000,"updateDate":1556532763000,"applyBackgroundUpdates":1,"path":"/tmp/rust_mozprofile.awdd2tQ6RVlm/extensions/injector@1.0.0","skinnable":false,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":true,"strictCompatibility":true,"locales":[],"targetApplications":[{"id":"toolkit@mozilla.org","minVersion":null,"maxVersion":null}],"targetPlatforms":[],"signedState":-1,"seen":true,"dependencies":[],"incognito":"spanning","userPermissions":{"permissions":[],"origins":["http://*/*","https://*/*"]},"icons":{},"iconURL":null,"blocklistState":0,"blocklistURL":null,"startupData":null,"hidden":false,"installTelemetryInfo":{"source":"app-profile","method":"sideload"},"rootURI":"file:///tmp/rust_mozprofile.awdd2tQ6RVlm/extensions/injector@1.0.0/","location":"app-profile"}
1556532773774	addons.xpi	DEBUG	Updating XPIState for {"id":"injector@1.0.0","syncGUID":"{bb2801d2-9ef5-44b4-a4a1-e4e3f5879bce}","version":"1.0.0","type":"extension","loader":null,"updateURL":null,"optionsURL":null,"optionsType":null,"optionsBrowserStyle":true,"aboutURL":null,"defaultLocale":{"name":"injector","creator":null,"developers":null,"translators":null,"contributors":null},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"installDate":1556532763000,"updateDate":1556532763000,"applyBackgroundUpdates":1,"path":"/tmp/rust_mozprofile.awdd2tQ6RVlm/extensions/injector@1.0.0","skinnable":false,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":true,"strictCompatibility":true,"locales":[],"targetApplications":[{"id":"toolkit@mozilla.org","minVersion":null,"maxVersion":null}],"targetPlatforms":[],"signedState":-1,"seen":true,"dependencies":[],"incognito":"spanning","userPermissions":{"permissions":[],"origins":["http://*/*","https://*/*"]},"icons":{},"iconURL":null,"blocklistState":0,"blocklistURL":null,"startupData":null,"hidden":false,"installTelemetryInfo":{"source":"app-profile","method":"sideload"},"rootURI":"file:///tmp/rust_mozprofile.awdd2tQ6RVlm/extensions/injector@1.0.0/","location":"app-profile"}
1556532773783	addons.xpi	DEBUG	Calling bootstrap method startup on injector@1.0.0 version 1.0.0
1556532774666	addons.xpi	DEBUG	Existing add-on injector@1.0.0 in app-profile
```

## Expected behaviour

The extension import should work as it does directly in firefox, either creating a temporary ID for the session, or honouring the addon ID in the manifest. And once imported, becoming active by default.

## Test script or set of commands reproducing this issue

manifest.json
```
{
  "manifest_version": 2,
  "name": "injector",
  "version": "1.0.0",
  "applications": {
    "gecko": {
      "id": "injector@injected.com"
    }
  },
  "content_scripts": [
    {
      "matches": [
        "http://*/*",
        "https://*/*"
      ],
      "js": [
        "injector.js"
      ]
    }
  ]
}
```

injector.js
```
document.body.innerHTML = "<h1>INJECTED!</h1>";
```

## Environment

OS: Ubuntu bionic 64 bit
Browser: Mozilla Firefox
Browser version: 67.0b2
Browser Driver version: GeckoDriver 0.24
Language Bindings version: ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]
Selenium Webdriver: 3.141.0 (latest version has unrelated fatal error)

**Comment by p0deje** [comment]
Thank you for the report! Have you checked in applying patch from #6550 solves the issue?

**Comment by theDogOfPavlov** [comment]
My pleasure.

I tried monkey-patching 3.141.0 but it fails with the same error:
```
1556543220757	addons.xpi	DEBUG	New add-on injector@1.0.0 in app-profile
1556543220788	addons.xpi-utils	DEBUG	New add-on injector@1.0.0 installed in app-profile
1556543220875	addons.xpi-utils	WARN	addMetadata: Add-on injector@1.0.0 is invalid: Error: Invalid addon ID: expected addon ID injector@1.0.0, found injector@injected.com in manifest(resource://gre/modules/addons/XPIDatabase.jsm:2357:15) JS Stack trace: addMetadata@XPIDatabase.jsm:2357:15
```
Maybe there is additional code to change ...

**Comment by p0deje** [comment]
Thanks for trying, I'll need to dig into this so and will do that in the next few days.

**Comment by theDogOfPavlov** [comment]
Thanks for the quick response!

**Comment by p0deje** [comment]
How do you add an extension? Are you use Firefox profile or calling `Driver#install_addon`? Can you please provide the code you use?

**Comment by theDogOfPavlov** [comment]
Of course!
```
profile = Selenium::WebDriver::Firefox::Profile.new
profile.add_extension( '/tmp/plugin' )
browser = Watir::Browser.new :firefox, profile: profile
```

**Comment by p0deje** [comment]
I think the issue is actually fixed in a25638dfb68e2752a224ab7716cb5edc491e4033. It still fails to install properly but now I see a different error:

```ruby
Selenium::WebDriver.logger.level = :debug
profile = Selenium::WebDriver::Firefox::Profile.new
profile.add_extension( '/tmp/plugin' )
options = Selenium::WebDriver::Firefox::Options.new(profile: profile)
Selenium::WebDriver.for(:firefox, options: options).quit
```

```
1556891014113	addons.xpi	DEBUG	New add-on injector@injected.com in app-profile
...
1556891014337	addons.xpi-utils	WARN	Add-on injector@injected.com is not correctly signed.
1556891014337	addons.xpi-utils	WARN	Add-on injector@injected.com is not correctly signed.
1556891014338	addons.xpi-utils	WARN	addMetadata: Add-on injector@injected.com is invalid: Error: Extension injector@injected.com is not correctly signed
```

Can you check if the fix works for you?

**Comment by theDogOfPavlov** [comment]
Yep, looks like that sorted it out.

The signing error you're now seeing is to be expected: you need the developers firefox build to run unsigned extensions.

Thanks for the quick turn-around!
